update contributing guide (#580)
fastplotlib/fastplotlib: CONTRIBUTING.md

2. Clone the repo. Replace the `YOUR_ACCOUNT` in the repo URL to the fork on your account.

> **_NOTE:_** We use [git-lfs](https://git-lfs.com) for storing large files, such as ground-truths for tests, so you will need 
> to [install it](https://github.com/git-lfs/git-lfs#installing) before cloning the repo.

```bash
git clone https://github.com/YOUR_ACCOUNT/fastplotlib.git
cd fastplotlib
```

3. Install `fastplotlib` in editable mode with developer dependencies 

```bash
# install all extras in place
pip install -e ".[notebook,docs,tests]"
```

> **_NOTE:_** If you cloned the repo before installing `git-lfs`, you can run `git lfs pull` at any
> time to download the files stored on LFS

4. Add the upstream remote branch:

```bash
git remote add upstream https://github.com/fastplotlib/fastplotlib
```

At this point you have two remotes: `origin` (your fork) and `upstream` (the canonical version). You won't have permission to push to upstream (only `origin`), but 
this make it easy to keep your `fastplotlib` up-to-date with the canonical version by pulling from upstream: `git pull upstream`.

### Creating a new branch

> As mentioned previously, each feature in `fastplotlib` is worked on in a separate branch. This allows multiple people developing multiple features simultaneously, without interfering with each other's work. To create

Reviewer: ```suggestion
As mentioned previously, each feature in `fastplotlib` is worked on in a separate branch. This allows multiple people to develop multiple features simultaneously, without interfering with each other's work. To create
```

documentation to properly note your changes.
See below for details on how to [add tests](#adding-tests) and properly [document](#adding-documentation) your code.

Now you are ready to make a Pull Request. You can open a pull request by clicking on the big `Compare & pull request` button that appears at the top of the `fastplotlib` repo 
after pushing to your branch (see [here](https://intersect-training.org/collaborative-git/03-pr/index.html) for a tutorial).

> **_NOTE:_** Please make sure that you initially make your PR as a **draft** PR against the `main` branch. When you think the PR is ready, mark
> it for review to trigger tests using our CI pipeline. If you need to make changes, please set the PR back to a draft when pushing further
> commits until it is ready for review again.

Your pull request should include the following:
- A summary including information on what you changed and why
- References to relevant issues or discussions
- Special notice to any portion of your changes where you have lingering questions (e.g., "was this the right way to implement this?") or
want reviewers to pay special attention to

> Next, we will be notified of the pull request and will read it over. We will try to give and initial response quickly, and then do a longer in-depth

Reviewer: ```suggestion
Next, we will be notified of the pull request and will read it over. We will try to give an initial response quickly, and then do a longer in-depth
```

review, at which point you will probably need to respond to our comments, making changes as appropriate. We will then respond again, and proceed
in an iterative fashion until everyone is happy with the proposed changes.

Once your changes are integrated, you will be added as a GitHub contributor. Thank you for being 

> apart of `fastplotlib`!

Reviewer: "a part"

### Style Guide

As far as code style, please adhere to the following guidelines:

- Longer, descriptive names are preferred (e.g., `x` is not an appropriate name for a variable), especially for anything user-facing,
such as methods, attributes, or arguments

> - Any public method or function must have complete type-annotated docstrings (see below for details). Hidden ones do not need to have

Reviewer: ```suggestion
- Any public method, property, or attribute must have complete type-annotated docstrings (see below for details). Private methods or attributes do not need to have a
```

complete docstring, but they probably should.

### Releases

> We create releases on GitHub, deploy on / distribute via [pypi](https://pypi.org/), and try to follow [semantic versioning](https://semver.org/):

Reviewer: ```suggestion
We create releases on GitHub and distribute via [pypi](https://pypi.org/), and try to follow [semantic versioning](https://semver.org/):
```

Reviewer: We're currently in alpha so we're not exactly following this, we're still at zero version

> Given a version number MAJOR.MINOR.PATCH, increment the:
> 1. MAJOR version when you make incompatible API changes
> 2. MINOR version when you add functionality in a backward compatible manner
> 3. PATCH version when you make backward compatible bug fixes

To release a new version, we [create a GitHub release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) with a new tag incrementing the version as described above. 
Creating the GitHub release will trigger the deployment to pypi, via our `deploy` action (found in `.github/workflows/pypi-publish.yml`). 

> The built version will grab the version tag from the GitHub release, using [setuptools_scm](https://github.com/pypa/setuptools_scm).

Reviewer: bro even I don't know this XD

Author: TYL

### Testing 

#### Testing Details

An integral part of our testing suite is not only to test that our code is working, but to also check that are plotting 

> library CI pipeline is producing things that "look visually correct". 

Reviewer: simplify, something like: "As a plotting library we require two layers of testing. We use a backend test suite that verifies the basic functionality of buffer managers, graphics, layouts, etc., and another test suite which verifies that the library renders plots that are _visually correct_"

In order to do this, each example within the `examples` directory is run and an image of the canvas is taken and compared 
with a ground-truth screenshot that we have manually inspected. Ground-truth images are stored using `git-lfs`.

The ground-truth images are located in:

```
examples/desktop/screenshots
examples/notebooks/screenshots
```

The tests will produce slightly different imperceptible (to a human) results on different hardware when compared to the 
ground-truth. A small RMSE tolerance has been chosen, `0.025` for most examples. If the output image and 
ground-truth image are within that tolerance the test will pass. 

Some feature development may require the ground-truth screenshots to be updated. In the event that your changes require
this, please do the following: 

1. Download the regenerated screenshots from the [`fastplotlib` GitHub Actions page](https://github.com/fastplotlib/fastplotlib/actions/workflows/screenshots.yml) for your specific PR

> 2. Replace the screenshots in your local `fastplotlib` screenshots directories with those downloaded 

Reviewer: replace only the screenshots that the PR changes

```
examples/desktop/screenshots
examples/notebooks/screenshots
```

3. Commit your new screenshots and push them to your branch to get picked up by `git-lfs`

```bash
# add changes
git add examples/desktop/screenshots/
git add examples/notebooks/screenshots/

# commit changes
git commit -m "update screenshots"

# push changes
git push origin my_feature_branch
```

#### Adding tests

> Depending on the type of contribution you are making, new tests will need to be added to the repository. Unit tests for testing underlying functionality such as buffer managers, figure instantiation, and

Reviewer: ```suggestion
Depending on the type of contribution you are making, new tests might need to be added to the repository. Unit tests for testing underlying functionality such as buffer managers, figure instantiation, and
```

more can be found in the `/tests` directory. However, we also test all of our desktop examples as well. 

> If you are adding a new example to the library, you will need to add to comments to the top of your `.py` file in order to make sure it is both tested and added to the gallery. 

Reviewer: ```suggestion
If you are adding a new example to the library, you will need the following comments to the top of your `.py` file in order to make sure it is both tested and added to the gallery. 
```

```python
# test_example = true
# sphinx_gallery_pygfx_docs = 'screenshot'
```

### Documentation

Documentation is a crucial part of open-source software and greatly influences the ability to use a codebase. As such, it is imperative that any new changes are
properly documented as outlined below. 

We use [`sphinx`](https://www.sphinx-doc.org/en/master/) for generating our documentation. In addition to this, we also use the [`sphinx-gallery`](https://sphinx-gallery.github.io/stable/index.html)
extension to build our examples gallery.

If you would like to build the documentation locally:

```bash
cd docs
# regenerate the api guide

> python source/generate_api.py

Reviewer: oh somewhere state that any changes to the API require this to be run before making a PR

Reviewer: Also any changes to Graphics constructors will require regenerating the mixin class, that info can go somewhere else :thinking:

# build locally
make html
```

#### Adding documentation

All public-facing functions and classes should have complete docstrings, which start with a one-line short summary of the function, 
a medium-length description of the function / class and what it does, and a complete description of all arguments and return values. 

> Docstrings should be relatively short, providing the information necessary for a user to use the code.

Reviewer: ```suggestion
Docstrings should comprehensive, providing the information necessary for a user to use the method or property without going through the code.
```

Our docstrings are  not short XD, good libs generally have comprehensive docstrings, see sklearn.
